keps/sig-instrumentation/4785-crdmetrics-controller/README.md

```diff
@@ -0,0 +1,1061 @@
+redeploy their modifications, event emissions on metrics generation, 
+multiple managed resources to isolate configuration logic, etc.
+
+#### Story 2
+
+As a cluster admin, I want to be able to express my Kube State Metrics' Custom
+Resource State configurations in a non-turing language that is well-known
+throughout the ecosystem, and enables me to get going without learning another
+DSL (domain-specific language).
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+The controller abides by the following principals:
+* Garbage in, garbage out: Invalid configurations will generate invalid metrics.
+  The exception to this being that certain checks that ensure metric structure
+  are still present (for e.g., `value` should be a `float64`).
+* Library support: The module is **not** intended to be used as a library, and 
+  as such, does not export any functions or types, with `pkg/` being an 
+  exception (for e.g., managed resource types).
+* Metrics stability: There are **no** metrics stability guarantees, as the
+  metrics are dynamically generated.
+* No middle-ware: The configuration is `unmarshal`led into a set of stores that
+  the codebase directly operates on. Unlike Kube State Metrics, there is **no**
+  controller-defined parsing middle-ware that processes the configuration before
+  it is used, in order to cut down on unnecessary complexity as much as
+  possible.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+N/A since the managed resource offered by the CRDMetrics controller provides the
+ability to define metric configurations that are a super-set of the 
+expressibility that Kube State Metrics' Custom Resource State configurations
+have to offer.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The controller offers a number of improvements over Kube State Metrics' Custom
+Resource State API, while maintaining a 3x faster round trip time for metric
+generation.
+
+- At its core, the controller relies on its managed resource,
+`CRDMetricsResource` to fetch the metric generation configuration. Parts of the
+configuration may be defined using different `resolver`s, such as `unstructured`
+or `CEL`.
```

Comment:
Is there a schema for this API kind?

Reply:
That's https://github.com/google/cel-go. It's [one of the configuration languages](https://github.com/rexagod/crdmetrics/tree/main/pkg/resolver) that the controller supports OOTB.

Reply:
I meant, can we see a schema for CRDMetricsResource?

Reply:
Sure, the type definitions are available [here](https://github.com/rexagod/crdmetrics/blob/main/pkg/apis/crdmetrics/v1alpha1/types.go), and the CRD schema is available [here](https://github.com/rexagod/crdmetrics/blob/main/manifests/custom-resource-definition.yaml).